File: ts/src/rsa4096/batch.ts
/**
 * Sequential generation of the first N (or any startIndex..startIndex+
 * count-1 range of) Arweave addresses from one seed bitstring, with ONE
 * combined progress stream across the whole run. Direct port of
 * RSA4096/batch.go -- see that file's doc comment for the full design
 * reasoning; this is a port, not a reinterpretation.
 *
 * This is orchestration on top of generateFromBitStringAtIndex/Async
 * (indexed.ts), not a new cryptographic mechanism -- every address is
 * still derived exactly the way indexed.ts already defines. What this
 * file adds is worth getting right ONCE, centrally: a single combined
 * 0..1 progress readout, results delivered incrementally as each
 * address completes, and one settled answer for partial failure (return
 * what completed so far, alongside the error, never silently discarded).
 *
 * Copyright (C) 2026 AncientHoldings GmbH. All rights reserved.
 */

import { generateFromBitStringAtIndex, generateFromBitStringAtIndexAsync } from './indexed.js';
import type { KeyGenResult } from './pipeline.js';
import type { ProgressCallback, ProgressStage } from './progress.js';

/**
 * Reported after every candidate draw of every address in the batch --
 * carries both that address's own detail and the combined, whole-batch
 * view. Matches RSA4096/batch.go's BatchProgressEvent exactly.
 */
export interface BatchProgressEvent {
  /** The absolute address index currently being searched for. */
  index: number;
  /** How many addresses in this batch have already finished. */
  completedCount: number;
  /** The batch's requested count, unchanged for the whole run. */
  totalCount: number;
  /** The CURRENT address's own overallProgress (see ProgressEvent) -- in [0, 1). */
  addressProgress: number;
  /**
   * Folds addressProgress into a single 0..1 value across the WHOLE
   * batch: (completedCount + addressProgress) / totalCount. Honest in
   * the same sense ProgressEvent.overallProgress is honest -- a real
   * weighted combination of N independent memoryless estimates, not a
   * fake animation.
   */
  overallProgress: number;
  /** Passed through from the current address's own ProgressEvent. */
  stage: ProgressStage;
  attempts: number;
}

/** May be omitted (undefined) -- zero overhead, exactly like ProgressCallback. */
export type BatchProgressCallback = (event: BatchProgressEvent) => void;

/**
 * Called once per completed address, in order, immediately as each one
 * finishes -- lets a caller render results progressively rather than
 * waiting for the whole batch. May be omitted.
 */
export type BatchResultCallback = (index: number, result: KeyGenResult) => void;

/** Thrown by both batch functions on invalid (startIndex, count) input. */
export class BatchRangeError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'BatchRangeError';
  }
}

/**
 * Thrown when a batch fails partway through. Unlike a plain Error, this
 * carries the results ALREADY completed (in order) so a caller doesn't
 * have to lose already-completed, multi-second-cost prime searches just
 * because a later index failed -- matches RSA4096/batch.go's "return
 * partial results alongside the error" semantics, which Go expresses via
 * a normal `([]*KeyGenResult, error)` return but TS expresses via this
 * dedicated error subclass carrying `completed`.
 */
export class BatchGenerationError extends Error {
  readonly completed: readonly KeyGenResult[];
  readonly failedIndex: number;
  readonly cause: unknown;

  constructor(
    message: string,
    completed: readonly KeyGenResult[],
    failedIndex: number,
    cause: unknown,
  ) {
    super(message);
    this.name = 'BatchGenerationError';
    this.completed = completed;
    this.failedIndex = failedIndex;
    this.cause = cause;
  }
}

const MAX_INDEX = 0xffffffff;

function validateRange(startIndex: number, count: number): void {
  if (!Number.isInteger(count) || count < 1) {
    throw new BatchRangeError(`count must be an integer >= 1, got ${count}`);
  }
  if (!Number.isInteger(startIndex) || startIndex < 0) {
    throw new BatchRangeError(`startIndex must be a non-negative integer, got ${startIndex}`);
  }
  if (startIndex + count - 1 > MAX_INDEX) {
    throw new BatchRangeError('startIndex + count - 1 exceeds the maximum index');
  }
}

/**
 * Synchronous batch generation -- see {@link generateBatchFromBitStringAsync}
 * for the event-loop-yielding variant a browser UI should use instead
 * (this variant blocks the calling thread for the ENTIRE batch, same
 * tradeoff as generateFromBitString vs generateFromBitStringAsync).
 *
 * startIndex === 0 with count === N gives "the first N addresses" -- the
 * primary use case -- but any starting point works (e.g. startIndex=100,
 * count=50 for "the next 50 after the first hundred").
 */
export function generateBatchFromBitString(
  seedBitString: string,
  startIndex: number,
  count: number,
  onProgress?: BatchProgressCallback,
  onResult?: BatchResultCallback,
): KeyGenResult[] {
  validateRange(startIndex, count);

  const results: KeyGenResult[] = [];
  for (let i = 0; i < count; i++) {
    const index = startIndex + i;
    const innerProgress: ProgressCallback | undefined = onProgress
      ? (ev) => {
          onProgress({
            index,
            completedCount: i,
            totalCount: count,
            addressProgress: ev.overallProgress,
            overallProgress: (i + ev.overallProgress) / count,
            stage: ev.stage,
            attempts: ev.attempts,
          });
        }
      : undefined;

    let result: KeyGenResult;
    try {
      result = generateFromBitStringAtIndex(seedBitString, index, innerProgress);
    } catch (cause) {
      throw new BatchGenerationError(
        `generateBatchFromBitString: failed at index ${index} (${i} of ${count} addresses completed before this): ${
          cause instanceof Error ? cause.message : String(cause)
        }`,
        results,
        index,
        cause,
      );
    }

    results.push(result);
    onResult?.(index, result);
  }

  return results;
}

/**
 * Async variant of {@link generateBatchFromBitString}, built on
 * {@link generateFromBitStringAtIndexAsync} -- yields to the event loop
 * during each address's prime search so a browser UI thread stays
 * responsive for the WHOLE batch, not just one address. This is the
 * recommended entry point for any UI that lets a user generate multiple
 * Arweave addresses from one seed.
 */
export async function generateBatchFromBitStringAsync(
  seedBitString: string,
  startIndex: number,
  count: number,
  onProgress?: BatchProgressCallback,
  onResult?: BatchResultCallback,
): Promise<KeyGenResult[]> {
  validateRange(startIndex, count);

  const results: KeyGenResult[] = [];
  for (let i = 0; i < count; i++) {
    const index = startIndex + i;
    const innerProgress: ProgressCallback | undefined = onProgress
      ? (ev) => {
          onProgress({
            index,
            completedCount: i,
            totalCount: count,
            addressProgress: ev.overallProgress,
            overallProgress: (i + ev.overallProgress) / count,
            stage: ev.stage,
            attempts: ev.attempts,
          });
        }
      : undefined;

    let result: KeyGenResult;
    try {
      result = await generateFromBitStringAtIndexAsync(seedBitString, index, innerProgress);
    } catch (cause) {
      throw new BatchGenerationError(
        `generateBatchFromBitStringAsync: failed at index ${index} (${i} of ${count} addresses completed before this): ${
          cause instanceof Error ? cause.message : String(cause)
        }`,
        results,
        index,
        cause,
      );
    }

    results.push(result);
    onResult?.(index, result);
  }

  return results;
}

File: RSA4096/batch.go
// File batch.go — sequential generation of the first N (or any
// startIndex..startIndex+count-1 range of) Arweave addresses from one
// seed bitstring, with ONE combined progress stream across the whole
// run.
//
// This is orchestration on top of GenerateFromBitStringAtIndex
// (indexed.go), not a new cryptographic mechanism — every address is
// still derived exactly the way indexed.go already defines. What this
// file adds is the part worth getting right ONCE, centrally, rather than
// leaving every caller (Codex, the StoicDigest demo, any future tool) to
// reimplement the same combined-progress arithmetic and error semantics
// slightly differently:
//
//   - a single 0..1 progress readout across the ENTIRE batch, not just
//     the address currently being searched for (see
//     BatchProgressEvent.OverallProgress);
//   - results delivered incrementally, one per completed address, via
//     onResult — so a caller can render address #1, #2, #3... as they
//     land instead of waiting for the whole batch to finish;
//   - one settled answer for "what happens if address 42 of 100 fails":
//     return the 41 already-completed results (not nil, not discarded)
//     ALONGSIDE the error, since each one cost real, multi-second work.
//
// Copyright (C) 2026 AncientHoldings GmbH. All rights reserved.
package RSA4096

import (
	"errors"
	"fmt"
	"math"
)

// BatchProgressEvent is reported after every candidate draw of every
// address in the batch (i.e. at the same frequency ProgressEvent already
// fires for a single address) — it carries both that address's own
// detail and the combined, whole-batch view.
type BatchProgressEvent struct {
	// Index is the absolute address index currently being searched for
	// (startIndex + CompletedCount, while it's in flight).
	Index uint32
	// CompletedCount is how many addresses in this batch have already
	// finished (0 while the first address is still being searched for).
	CompletedCount uint32
	// TotalCount is the batch's requested count, unchanged for the
	// whole run — convenience so a caller doesn't have to keep it
	// separately.
	TotalCount uint32
	// AddressProgress is the CURRENT address's own OverallProgress
	// (see ProgressEvent) — in [0, 1).
	AddressProgress float64
	// OverallProgress folds AddressProgress into a single 0..1 value
	// across the WHOLE batch: (CompletedCount + AddressProgress) /
	// TotalCount. Honest in the same sense ProgressEvent.OverallProgress
	// is honest — a real weighted combination of N independent
	// memoryless estimates, not a fake animation.
	OverallProgress float64
	// Stage and Attempts pass through the current address's own
	// ProgressEvent detail, for a caller that wants both a fine-grained
	// per-address bar and the combined one.
	Stage    ProgressStage
	Attempts int
}

// BatchProgressFunc receives one BatchProgressEvent per candidate draw,
// across every address in the batch. May be nil (zero overhead, exactly
// like ProgressFunc).
type BatchProgressFunc func(BatchProgressEvent)

// BatchResultFunc is called once per completed address, in order,
// immediately as each one finishes — lets a caller render results
// progressively rather than waiting for the whole batch. May be nil.
type BatchResultFunc func(index uint32, result *KeyGenResult)

// GenerateBatchFromBitString derives `count` independent Arweave
// addresses — indices startIndex, startIndex+1, ..., startIndex+count-1
// — from the same seed bitstring, sequentially (not in parallel — each
// address's prime search is CPU-bound and independent, so there is no
// benefit to interleaving them, and sequential execution is what makes
// "one combined progress bar" a coherent, linearly-advancing thing to
// show), reporting one COMBINED progress stream across the whole run.
//
// startIndex == 0 with count == N gives "the first N addresses" — the
// primary use case — but any starting point is directly usable (e.g.
// startIndex=100, count=50 for "the next 50 after the first hundred"),
// consistent with indexed.go's "any index directly reachable" design.
//
// On error partway through, returns the results completed SO FAR (in
// order, one entry per completed index, length == the number completed)
// ALONGSIDE the error — never nil-with-error, since each completed
// result cost real, multi-second work that a caller should not have to
// discard just because a later index failed.
func GenerateBatchFromBitString(
	seedBitString string,
	startIndex, count uint32,
	onProgress BatchProgressFunc,
	onResult BatchResultFunc,
) ([]*KeyGenResult, error) {
	if count == 0 {
		return nil, errors.New("GenerateBatchFromBitString: count must be at least 1")
	}
	if uint64(startIndex)+uint64(count)-1 > uint64(math.MaxUint32) {
		return nil, errors.New("GenerateBatchFromBitString: startIndex + count - 1 exceeds the maximum index")
	}

	results := make([]*KeyGenResult, 0, count)
	for i := uint32(0); i < count; i++ {
		idx := startIndex + i

		var innerProgress ProgressFunc
		if onProgress != nil {
			completedSoFar := i // captured by value per-iteration (Go 1.22+ loop var semantics; see go.mod)
			innerProgress = func(ev ProgressEvent) {
				overall := (float64(completedSoFar) + ev.OverallProgress) / float64(count)
				onProgress(BatchProgressEvent{
					Index:           idx,
					CompletedCount:  completedSoFar,
					TotalCount:      count,
					AddressProgress: ev.OverallProgress,
					OverallProgress: overall,
					Stage:           ev.Stage,
					Attempts:        ev.Attempts,
				})
			}
		}

		result, err := GenerateFromBitStringAtIndex(seedBitString, idx, innerProgress)
		if err != nil {
			return results, fmt.Errorf(
				"GenerateBatchFromBitString: failed at index %d (%d of %d addresses completed before this): %w",
				idx, i, count, err,
			)
		}

		results = append(results, result)
		if onResult != nil {
			onResult(idx, result)
		}
	}

	return results, nil
}
